### ISSUE TITLE
Regression(?) in LSTM behavior from 2.2.5 to 2.3.1

### ISSUE THREAD

**Issue by kylrth** [keras-team/keras#13869]
I'm noticing a change in behavior between 2.2.5 and 2.3.1 with the LSTM layer.

**System information**  
- OS Platform and Distribution: Ubuntu 18.04 Docker container
- TensorFlow backend
- Python version: 3.6.9
- CPU-only

**Code to reproduce the issue**

```python
from keras.layers import LSTM, Dense, Input, Lambda

layer = LSTM(
    256,
    activation="tanh",
    recurrent_activation="sigmoid",
    kernel_initializer="he_normal",
    recurrent_initializer="orthogonal",
    return_sequences=True,
    name="03_lstm",
    activity_regularizer=None,
    kernel_regularizer=None,
    bias_regularizer=None,
    recurrent_dropout=0.0,
    return_state=True,
)

i_state = Input(shape=(2, 256), name="input_state_03")

state_h0 = Lambda(lambda x: x[:, 0:1, :], name="03_init_state_h")(i_state)
state_c0 = Lambda(lambda x: x[:, 1:2, :], name="03_init_state_c")(i_state)

h = Input(shape=(None, 513), name="input")
h = Dense(
    256,
    activation="tanh",
    kernel_initializer="he_normal",
    name="03_dense",
    activity_regularizer=None,
    kernel_regularizer=None,
    bias_regularizer=None,
)(h)

h, state_h, state_c = layer(h, initial_state=[state_h0, state_c0])
```

#### TF 1.14.0 and Keras 2.2.5

Runs without error.

#### TF 1.14.0 and Keras 2.3.1

```
Traceback (most recent call last):
  File "/usr/local/lib/python3.6/dist-packages/tensorflow/python/framework/ops.py", line 1864, in _create_c_op
    c_op = c_api.TF_FinishOperation(op_desc)
tensorflow.python.framework.errors_impl.InvalidArgumentError: Dimensions must be equal, but are 1024 and 256 for '03_lstm/mul' (op: 'Mul') with input shapes: [?,?,1024], [?,1,256].

During handling of the above exception, another exception occurred:

Traceback (most recent call last):
  File "keras_test.py", line 34, in <module>
    h, state_h, state_c = layer(h, initial_state=[state_h0, state_c0])
  File "/usr/local/lib/python3.6/dist-packages/keras/layers/recurrent.py", line 584, in __call__
    output = super(RNN, self).__call__(full_input, **kwargs)
  File "/usr/local/lib/python3.6/dist-packages/keras/engine/base_layer.py", line 489, in __call__
    output = self.call(inputs, **kwargs)
  File "/usr/local/lib/python3.6/dist-packages/keras/layers/recurrent.py", line 2247, in call
    initial_state=initial_state)
  File "/usr/local/lib/python3.6/dist-packages/keras/layers/recurrent.py", line 682, in call
    input_length=timesteps)
  File "/usr/local/lib/python3.6/dist-packages/keras/backend/tensorflow_backend.py", line 3103, in rnn
    input_length=input_length)
  File "/usr/local/lib/python3.6/dist-packages/tensorflow/python/keras/backend.py", line 3732, in rnn
    input_time_zero, tuple(initial_states) + tuple(constants))
  File "/usr/local/lib/python3.6/dist-packages/keras/layers/recurrent.py", line 673, in step
    return self.cell.call(inputs, states, **kwargs)
  File "/usr/local/lib/python3.6/dist-packages/keras/layers/recurrent.py", line 2049, in call
    c = f * c_tm1 + i * self.activation(z2)
  File "/usr/local/lib/python3.6/dist-packages/tensorflow/python/ops/math_ops.py", line 884, in binary_op_wrapper
    return func(x, y, name=name)
  File "/usr/local/lib/python3.6/dist-packages/tensorflow/python/ops/math_ops.py", line 1180, in _mul_dispatch
    return gen_math_ops.mul(x, y, name=name)
  File "/usr/local/lib/python3.6/dist-packages/tensorflow/python/ops/gen_math_ops.py", line 6490, in mul
    "Mul", x=x, y=y, name=name)
  File "/usr/local/lib/python3.6/dist-packages/tensorflow/python/framework/op_def_library.py", line 788, in _apply_op_helper
    op_def=op_def)
  File "/usr/local/lib/python3.6/dist-packages/tensorflow/python/util/deprecation.py", line 507, in new_func
    return func(*args, **kwargs)
  File "/usr/local/lib/python3.6/dist-packages/tensorflow/python/framework/ops.py", line 3616, in create_op
    op_def=op_def)
  File "/usr/local/lib/python3.6/dist-packages/tensorflow/python/framework/ops.py", line 2027, in __init__
    control_input_ops)
  File "/usr/local/lib/python3.6/dist-packages/tensorflow/python/framework/ops.py", line 1867, in _create_c_op
    raise ValueError(str(e))
ValueError: Dimensions must be equal, but are 1024 and 256 for '03_lstm/mul' (op: 'Mul') with input shapes: [?,?,1024], [?,1,256].
ERROR:tensorflow:==================================
Object was never used (type <class 'tensorflow.python.ops.tensor_array_ops.TensorArray'>):
<tensorflow.python.ops.tensor_array_ops.TensorArray object at 0x7f1756394b38>
If you want to mark it as used call its "mark_used()" method.
It was originally created here:
  File "/usr/local/lib/python3.6/dist-packages/tensorflow/python/keras/backend.py", line 3722, in <genexpr>
    for ta, input_ in zip(input_ta, flatted_inputs))  File "/usr/local/lib/python3.6/dist-packages/tensorflow/python/util/tf_should_use.py", line 193, in wrapped
    return _add_should_use_warning(fn(*args, **kwargs))
==================================
ERROR:tensorflow:==================================
Object was never used (type <class 'tensorflow.python.ops.tensor_array_ops.TensorArray'>):
<tensorflow.python.ops.tensor_array_ops.TensorArray object at 0x7f17563948d0>
If you want to mark it as used call its "mark_used()" method.
It was originally created here:
  File "/usr/local/lib/python3.6/dist-packages/tensorflow/python/keras/backend.py", line 3722, in <genexpr>
    for ta, input_ in zip(input_ta, flatted_inputs))  File "/usr/local/lib/python3.6/dist-packages/tensorflow/python/util/tf_should_use.py", line 193, in wrapped
    return _add_should_use_warning(fn(*args, **kwargs))  File "/usr/local/lib/python3.6/dist-packages/tensorflow/python/ops/tensor_array_ops.py", line 1256, in unstack
    return self._implementation.unstack(value, name=name)  File "/usr/local/lib/python3.6/dist-packages/tensorflow/python/util/tf_should_use.py", line 193, in wrapped
    return _add_should_use_warning(fn(*args, **kwargs))
==================================
ERROR:tensorflow:==================================
Object was never used (type <class 'tensorflow.python.ops.tensor_array_ops.TensorArray'>):
<tensorflow.python.ops.tensor_array_ops.TensorArray object at 0x7f1756394e48>
If you want to mark it as used call its "mark_used()" method.
It was originally created here:
  File "/usr/local/lib/python3.6/dist-packages/tensorflow/python/keras/backend.py", line 3722, in <genexpr>
    for ta, input_ in zip(input_ta, flatted_inputs))  File "/usr/local/lib/python3.6/dist-packages/tensorflow/python/util/tf_should_use.py", line 193, in wrapped
    return _add_should_use_warning(fn(*args, **kwargs))  File "/usr/local/lib/python3.6/dist-packages/tensorflow/python/ops/tensor_array_ops.py", line 1256, in unstack
    return self._implementation.unstack(value, name=name)  File "/usr/local/lib/python3.6/dist-packages/tensorflow/python/util/tf_should_use.py", line 193, in wrapped
    return _add_should_use_warning(fn(*args, **kwargs))  File "/usr/local/lib/python3.6/dist-packages/tensorflow/python/ops/tensor_array_ops.py", line 351, in unstack
    indices=math_ops.range(0, num_elements), value=value, name=name)  File "/usr/local/lib/python3.6/dist-packages/tensorflow/python/util/tf_should_use.py", line 193, in wrapped
    return _add_should_use_warning(fn(*args, **kwargs))
==================================
```

#### TF 2.1.0 and Keras 2.3.1

```
Traceback (most recent call last):
  File "/usr/local/lib/python3.6/dist-packages/tensorflow_core/python/framework/ops.py", line 1619, in _create_c_op
    c_op = c_api.TF_FinishOperation(op_desc)
tensorflow.python.framework.errors_impl.InvalidArgumentError: Dimensions must be equal, but are 1024 and 256 for '03_lstm/mul' (op: 'Mul') with input shapes: [?,?,1024], [?,1,256].

During handling of the above exception, another exception occurred:

Traceback (most recent call last):
  File "keras_test.py", line 34, in <module>
    h, state_h, state_c = layer(h, initial_state=[state_h0, state_c0])
  File "/usr/local/lib/python3.6/dist-packages/keras/layers/recurrent.py", line 584, in __call__
    output = super(RNN, self).__call__(full_input, **kwargs)
  File "/usr/local/lib/python3.6/dist-packages/keras/backend/tensorflow_backend.py", line 75, in symbolic_fn_wrapper
    return func(*args, **kwargs)
  File "/usr/local/lib/python3.6/dist-packages/keras/engine/base_layer.py", line 489, in __call__
    output = self.call(inputs, **kwargs)
  File "/usr/local/lib/python3.6/dist-packages/keras/layers/recurrent.py", line 2247, in call
    initial_state=initial_state)
  File "/usr/local/lib/python3.6/dist-packages/keras/layers/recurrent.py", line 682, in call
    input_length=timesteps)
  File "/usr/local/lib/python3.6/dist-packages/keras/backend/tensorflow_backend.py", line 3103, in rnn
    input_length=input_length)
  File "/usr/local/lib/python3.6/dist-packages/tensorflow_core/python/keras/backend.py", line 4027, in rnn
    input_time_zero, tuple(initial_states) + tuple(constants))
  File "/usr/local/lib/python3.6/dist-packages/keras/layers/recurrent.py", line 673, in step
    return self.cell.call(inputs, states, **kwargs)
  File "/usr/local/lib/python3.6/dist-packages/keras/layers/recurrent.py", line 2049, in call
    c = f * c_tm1 + i * self.activation(z2)
  File "/usr/local/lib/python3.6/dist-packages/tensorflow_core/python/ops/math_ops.py", line 902, in binary_op_wrapper
    return func(x, y, name=name)
  File "/usr/local/lib/python3.6/dist-packages/tensorflow_core/python/ops/math_ops.py", line 1201, in _mul_dispatch
    return gen_math_ops.mul(x, y, name=name)
  File "/usr/local/lib/python3.6/dist-packages/tensorflow_core/python/ops/gen_math_ops.py", line 6125, in mul
    "Mul", x=x, y=y, name=name)
  File "/usr/local/lib/python3.6/dist-packages/tensorflow_core/python/framework/op_def_library.py", line 742, in _apply_op_helper
    attrs=attr_protos, op_def=op_def)
  File "/usr/local/lib/python3.6/dist-packages/tensorflow_core/python/framework/func_graph.py", line 595, in _create_op_internal
    compute_device)
  File "/usr/local/lib/python3.6/dist-packages/tensorflow_core/python/framework/ops.py", line 3322, in _create_op_internal
    op_def=op_def)
  File "/usr/local/lib/python3.6/dist-packages/tensorflow_core/python/framework/ops.py", line 1786, in __init__
    control_input_ops)
  File "/usr/local/lib/python3.6/dist-packages/tensorflow_core/python/framework/ops.py", line 1622, in _create_c_op
    raise ValueError(str(e))
ValueError: Dimensions must be equal, but are 1024 and 256 for '03_lstm/mul' (op: 'Mul') with input shapes: [?,?,1024], [?,1,256].
```

What is causing the tensor to have shape `[?,?,1024]` in Keras 2.3.1, but `[?,?,256]` in Keras 2.2.5?

**Comment by RasmusD** [comment]
I am encountering the same error on a similar system with the same differences between Keras and TF versions. Would be nice to get fixed or figured out what may be wrong as it currently blocks me from upgrading to TF 2.x use.

**Comment by shahruk10** [comment]
I face the same issue. The error message is quite obtuse; the actual problem turned out to be a mismatch in the expected shape of the tensors for the initial states of the LSTM layer. 

Tensorflow-1.15 onwards, the initial states are expected to be flat tensors (1-D). So for an LSTM layer with 256 units, both the hidden and cell states are expected to be have shape =  `(None, 256)` where None is just a place holder for the batch size.

As you are passing the states in via a 2-D tensor input layer with shape `(None, 2, 256)`, and extracting the hidden and cell states, you end up with tensors of shape `(None, 1, 256)` for `state_h0` and `state_c0`. Pre tensorflow-1.15, this would also be accepted by the LSTM layer and probably be flattened internally (I haven't checked how it's handled).

##### TLDR:

Using `tf.squeeze` or even a `Flatten` layer to remove the redundant dimension will solve the issue:

```diff
+from keras.layers import Flatten

+state_h0 = Flatten()(state_h0)
+state_c0 = Flatten()(state_c0)
h, state_h, state_c = layer(h, initial_state=[state_h0, state_c0])
```

**Comment by RasmusD** [comment]
Excellent! Good to know!

**Comment by sushreebarsa** [comment]
@kylrth Moving this issue to closed status as there has been no recent activity, in case you still face the error please create a new issue.Thanks!
